### DISCUSSION TITLE
How are you supposed to use relations? (switching from a php doctrine/orm)

### DISCUSSION THREAD

**Discussion by Nemo64** [prisma/prisma#20837] [Q&A]
### Question

Hi,

I'm currently trying to use prisma in a next application as a replacement for a way more complicated setup.
Atm., I used to use a symfony application with a doctrine ORM using an API generator for data storage.
I thought that I could replace all that by just using the database directly with prisma.

However: I'm not quite sure how to build any complex model with it.
In any application, you'd like to have a model that has grouped properties.
There are composite types in prisma (which would sometimes solve it) but only for mongodb.
Alternatively, I could model embedded objects with relations (and sometimes you have to).
Relations are easily defined in prisma (which is good) but you always get the reference ID. adding `@ignore` to it doesn't work.
```prisma
  project   Project  @relation(references: [id], onDelete: Cascade, fields: [projectId])
  projectId Int      @ignore // this breaks
```
So you have to fetch the reference ID. Ok, unnecessary, but not that wild. Just clutters the object model.
But then I try to write the relation, and now I have to manually remove the `projectId` because I obviously can't write it.

It gets even worse with `one-to-many` relations. A model like this is easily defined:
```prisma
model Project {
  id          Int                @id @default(autoincrement())
  name        String
  timeFrames  ProjectTimeFrame[]
}

model ProjectTimeFrame {
  id        Int      @id @default(autoincrement())
  project   Project  @relation(references: [id], onDelete: Cascade, fields: [projectId])
  projectId Int
}
```

But if I try to use it, I'll have to write an update query like this:

```typescript
prisma.project.update({
  where: { id: project.id },
  include: { timeFrames: true },
  data: {
    ...values,
    timeFrames: toManyUpdate(
      values.timeFrames.map(({ projectId, ...rest }) => rest), // remove the projectId
    ),
  },
});

// [...]

function toManyUpdate<T extends { id?: number }>(records: T[]) {
  const oldRecords = records.filter(
    (record): record is T & { id: number } => record.id !== undefined,
  );
  const newRecords = records.filter(
    (record): record is T & { id: undefined } => record.id === undefined,
  );
  return {
    deleteMany: { id: { notIn: oldRecords.map(({ id }) => id) } },
    update: oldRecords.map(({ id, ...data }) => ({ where: { id }, data })),
    create: newRecords,
  };
}
```

It all appears just way too complicated for such a simple things.

Am I completely on the wrong track?

**Comment by nurul3101** [comment]
Hello @Nemo64 👋 

Thank you for opening a GitHub Discussion and providing us with all the details.

> I'm currently trying to use prisma in a next application as a replacement for a way more complicated setup.
Atm., I used to use a symfony application with a doctrine ORM using an API generator for data storage.
I thought that I could replace all that by just using the database directly with prisma.

Do you want to migrate an existing application to Next.js with Prisma or are you starting the application from scratch?
If you are migrating an existing application then you can [introspect your database](https://www.prisma.io/docs/getting-started/setup-prisma/add-to-existing-project/relational-databases-typescript-postgresql) to generate a `schema.prisma` file automatically.

> In any application, you'd like to have a model that has grouped properties.
There are composite types in prisma (which would sometimes solve it) but only for mongodb.

Indeed, composite types are only supported for MongoDB. However, you can use [JSON types](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-json-fields) to simulate the behaviour of composite types in the relational database. The storage of the grouped properties will be more efficient compared to defining a relation, as data is physically stored in the same table.

> Alternatively, I could model embedded objects with relations (and sometimes you have to).

Modelling relations via foreign keys and normalising data are indeed the default way to express relationships between tables in a database.

> Relations are easily defined in prisma (which is good) but you always get the reference ID. adding @ignore to it doesn't work.

Do you mind elaborating on what you mean by "but you always get the reference ID"?
Do you mean you don't want to define the reference ID ( i.e. foreign key) in the `schema.prisma` file, or do you want to exclude the reference ID from the query response? If it is the latter, then you can use [`select`](https://www.prisma.io/docs/concepts/components/prisma-client/select-fields) to exclude it from the response. 

Based in the query that you shared:
```ts
prisma.project.update({
  where: { id: project.id },
  include: { timeFrames: true },
  data: {
    ...values,
    timeFrames: toManyUpdate(
      values.timeFrames.map(({ projectId, ...rest }) => rest), // remove the projectId
    ),
  },
});
```

It looks like you want to update `ProjectTimeFrame` records. Is there any particular reason why you are updating with a [nested write](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-writes) with the `Project` model? You could simply use the `ProjectTimeFrame` model to update the records like this:

```ts
prisma. projectTimeFrame.update({
  where: { projectId: project.id },
  data: {
    // Your existing logic here
  },
});
```

You might also want to have a look at [`updateMany`](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#updatemany) which would allow you to update multiple `ProjectTimeFrame` records at once.

Please let us know in case you have any questions. Hope this helps!

It would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 
Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by Nemo64** [reply]
Thanks for the answer @nurul3101,

> Do you want to migrate an existing application to Next.js with Prisma or are you starting the application from scratch?

I'm currently in the evaluating if prisma is a good fit for the projects we do at our agency so i appreciate your perspective.
My test project is created from scratch. 

> Indeed, composite types are only supported for MongoDB. However, you can use [JSON types](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-json-fields) to simulate the behaviour of composite types in the relational database.

The JSON type has the large disadvantage of not actually modeling fields in the database. Dcotrine (the php orm I currently use) has something called [Embeddables](https://www.doctrine-project.org/projects/doctrine-orm/en/2.9/tutorials/embeddables.html), where each field is mapped to an actual field in SQL. You can use them to search, you can set indexes on them, you can put relations in them and you can quickly select a group of parameters, that belong together, with a single mention of that embedded field.

It's great for organizing larger models.

> Do you mind elaborating on what you mean by "but you always get the reference ID"?

You guessed right. I mean the foreign key. In my case, the `projectId`. That is an implementation detail within the database but since I try to treat the `Project` with it's `ProjectTimeFrames` as a single model, it's irrelevant for the client ~ usually.

> It looks like you want to update ProjectTimeFrame records. Is there any particular reason why you are updating with a [nested write](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-writes) with the Project model?

There is. I'm not just updating the nested records. I'm using nested records as part of the project entity. Think of a single form, that shows the project name and all it's time frames. Thats what I'm building. That kind of relation is pretty common tasks for me.
If I could, I'd really like to model and treat it like a single Model, just like it is possible with mongodb. But with the added querying and indexing flexibility of an SQL database. 

And I don't know the user input in advance, so it could be that some time frames are deleted, some are updated, and some are added. That's why I build that complex `toManyUpdate` function. This also means that updating them separatly wouldn't work (and it would also not be in a single transaction, which is also not great)

I would have assumed that those kind of CRUD operations are pretty common, which is why it's curious to me that there is no direct handling for it. I first thought that `timeFrames: { set: project.timeFrames }` could be the answer but that operation is not meant to update nested records, if I understand correctly.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
